types/babel__traverse/index.d.ts

```diff
@@ -299,7 +299,7 @@ export class Binding {
 export type Visitor<S = unknown> =
     & VisitNodeObject<S, Node>
     & {
-        [Type in Node["type"]]?: VisitNode<S, Extract<Node, { type: Type }>>;
+        [N in Node as N["type"]]?: VisitNode<S, N extends { type: N["type"] } ? N : never>;
```

Comment:
I can think of a couple of reasons; one is just that you've switched (I think?) the mapped type to be homomorphic by specifically doing `N in Node as N["type"]`; this may actually end up being a behavior change just due to how these mappings work...

Reply:
No obvious outward differences jumped out at me between the two kinds of mapped types at the bottom of [this little playground](https://tsplay.dev/WY1R2W) (albeit it's hard to see the whole picture) besides the aforementioned speed. Can you think of what the downstream behavioural changes may be? If so I'm happy to investigate a bit further.

Reply:
I'll be honest and say that I'm not an expert here; @weswigham would probably know how to better reason about this as I'm more or less parroting off of others 😄

Reply:
`Extract` is distributive, so operates over every member of the `Node` passed in - so for every key in `Node["type"]` previously it was iterating over every `Node` type to find and select the desired type (exponential behavior). Now, instead, for each member of `Node`, we only create a single property and execute a single conditional over that member. We're just iterating over less.